bioio_ome_zarr/writers/ome_zarr_writer_v3.py

```diff
@@ -0,0 +1,541 @@
+
+    crop_slices: List[Union[slice, int]] = []
+    reshape_dims: List[int] = []
+    factor_axes: List[int] = []
+    new_dim_idx = 0
+
+    for i, dim_size in enumerate(shape):
+        ax_name = names_lower[i]
+        f = scale_factors[i]
+
+        if ax_name in ("x", "y") and f > 1 and dim_size >= f:
+            full = (dim_size // f) * f
+            crop_slices.append(slice(0, full))
+            reshape_dims.extend([full // f, f])
+            factor_axes.append(new_dim_idx + 1)
+            new_dim_idx += 2
+        else:
+            crop_slices.append(slice(None))
+            reshape_dims.append(dim_size)
+            new_dim_idx += 1
+
+    data_cropped = data[tuple(crop_slices)]
+    data_reshaped = data_cropped.reshape(tuple(reshape_dims))
+    downsampled = data_reshaped.mean(axis=tuple(factor_axes))
+
+    return downsampled
+
+
+class OMEZarrWriterV3:
+    """
+    OMEZarrWriterV3 is a fully compliant OME-Zarr v0.5.0 writer built
+    on Zarr v3 stores. Supports 2 ≤ N ≤ 5 dimensions (e.g. YX, ZYX,
+    TYX, CZYX, or TCZYX).
+    """
+
+    def __init__(
+        self,
+        store: Union[str, zarr.storage.StoreLike],
+        shape: Tuple[int, ...],
+        dtype: Union[np.dtype, str],
+        axes_names: Optional[List[str]] = None,
+        axes_types: Optional[List[str]] = None,
+        axes_units: Optional[List[Optional[str]]] = None,
+        axes_scale: Optional[List[float]] = None,
+        scale_factors: Optional[Tuple[int, ...]] = None,
+        num_levels: Optional[int] = None,
+        chunks: Union[str, Tuple[int, ...], List[Tuple[int, ...]]] = "auto",
+        shards: Optional[Union[Tuple[int, ...], List[Tuple[int, ...]]]] = None,
+        compressor: Optional[BloscCodec] = None,
+        image_name: str = "Image",
+        channels: Optional[List[Channel]] = None,
+        rdefs: Optional[dict] = None,
+        creator_info: Optional[dict] = None,
+        multiscale_scale: Optional[List[float]] = None,
+    ):
+        """
+        Initialize writer and build axes + channel metadata automatically.
+
+        Parameters
+        ----------
+        store : Union[str, zarr.storage.StoreLike]
+            Path or Zarr store-like object for the output group.
+        shape : Tuple[int, ...]
+            Base image shape (e.g. (Y, X), (Z, Y, X), (T, C, Z, Y, X)).
+        dtype : Union[np.dtype, str]
+            NumPy dtype of the image data (e.g. "uint8").
+        axes_names : Optional[List[str]]
+            Names of each axis, len = len(shape). Defaults to last N of
+            ["t", "c", "z", "y", "x"].
+        axes_types : Optional[List[str]]
+            Types of each axis (e.g. ["time", "channel", "space"...]).
+        axes_units : Optional[List[Optional[str]]]
+            Physical units for each axis (e.g. ["ms", None, "µm"]).
+        axes_scale : Optional[List[float]]
+            Physical scale per axis at base resolution.
+        scale_factors : Optional[Tuple[int, ...]]
```

Comment:
it would be a good future extension to take this as a list of tuples so that you don't have to downsample by the exact same factor at each level.  For instance, one could precompute the desired level shapes and then calculate the scale_factors at each level just by taking the ratios of the shapes.